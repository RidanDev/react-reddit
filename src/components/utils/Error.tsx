import React, { FunctionComponent } from 'react';
import { Theme } from '@material-ui/core';
import { createStyles, makeStyles } from '@material-ui/styles';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import ReportIcon from '@material-ui/icons/Report';
import classNames from 'classnames';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    error: {
      display: 'flex',
      flexGrow: 1,
      alignItems: 'center',
      justifyContent: 'center',
      flexDirection: 'column',
    },
    gutters: {
      ...theme.mixins.gutters(),
      paddingTop: theme.spacing.unit * 2,
      paddingBottom: theme.spacing.unit * 2,
    },
    text: {
      marginTop: theme.spacing.unit,
      marginBottom: theme.spacing.unit,
    },
  }),
);

export interface IErrorProps {
  message?: string;
  buttonMessage?: string;
  gutters?: boolean;
  onButtonClick?(event: React.MouseEvent<HTMLElement>): void;
}

const Error: FunctionComponent<IErrorProps> = React.memo(
  ({ onButtonClick, message, buttonMessage, gutters = false }) => {
    const classes = useStyles();

    return (
      <div
        data-testid="error"
        className={classNames(classes.error, {
          [classes.gutters]: gutters,
        })}
      >
        <ReportIcon style={{ fontSize: 80 }} color="disabled" />
        <Typography variant="body1" className={classes.text}>
          {message || 'Oops, something went wrong'}
        </Typography>
        <Button color="primary" onClick={onButtonClick}>
          {buttonMessage || 'Retry'}
        </Button>
      </div>
    );
  },
);

export default Error;
