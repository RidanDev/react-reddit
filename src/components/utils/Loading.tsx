import React, { FunctionComponent } from 'react';
import { createStyles, makeStyles } from '@material-ui/styles';
import CircularProgress from '@material-ui/core/CircularProgress';

const useStyles = makeStyles(
  createStyles({
    loading: {
      display: 'flex',
      flexGrow: 1,
      alignItems: 'center',
      justifyContent: 'center',
    },
  }),
);

const Loading: FunctionComponent<{}> = React.memo(() => {
  const classes = useStyles();

  return (
    <div className={classes.loading}>
      <CircularProgress />
    </div>
  );
});

export default Loading;
