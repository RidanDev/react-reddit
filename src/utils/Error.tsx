import React, { FunctionComponent } from 'react';
import { createStyles, withStyles, WithStyles, Theme } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import ReportIcon from '@material-ui/icons/Report';

const styles = (theme: Theme) =>
  createStyles({
    error: {
      display: 'flex',
      flexGrow: 1,
      alignItems: 'center',
      justifyContent: 'center',
      flexDirection: 'column',
    },
    text: {
      marginTop: theme.spacing.unit,
      marginBottom: theme.spacing.unit,
    },
  });

export interface IErrorProps extends WithStyles<typeof styles> {
  onRetryClicked?(): void;
}

const Error: FunctionComponent<IErrorProps> = ({ classes, onRetryClicked }) => {
  return (
    <div className={classes.error}>
      <ReportIcon style={{ fontSize: 80 }} color="disabled" />
      <Typography variant="body1" className={classes.text}>
        Oops, something went wrong
      </Typography>
      <Button color="primary" onClick={onRetryClicked}>
        Retry
      </Button>
    </div>
  );
};

export default withStyles(styles)(Error);
