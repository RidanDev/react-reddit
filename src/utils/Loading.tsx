import React, { FunctionComponent } from 'react';
import { createStyles, withStyles, WithStyles } from '@material-ui/core/styles';
import CircularProgress from '@material-ui/core/CircularProgress';

const styles = () =>
  createStyles({
    loading: {
      display: 'flex',
      flexGrow: 1,
      alignItems: 'center',
      justifyContent: 'center',
    },
  });

export interface ILoadingProps extends WithStyles<typeof styles> {}

const Loading: FunctionComponent<ILoadingProps> = React.memo(({ classes }) => {
  return (
    <div className={classes.loading}>
      <CircularProgress />
    </div>
  );
});

export default withStyles(styles)(Loading);
