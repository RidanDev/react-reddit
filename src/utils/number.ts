const mapping = [
  { number: 1000000000, postfix: 'b' },
  { number: 1000000, postfix: 'm' },
  { number: 1000, postfix: 'k' },
];

export interface IHumanizeOptions {
  /**
   * Adds a space between the number and the letter
   */
  space?: boolean;
  /**
   * Transform the letter to uppercase
   */
  upperCase?: boolean;
  /**
   * The number of decimal places that have to be preserved
   */
  precision?: number;
}

export function humanize(
  num: number,
  { space, upperCase, precision = 0 }: IHumanizeOptions = {},
): string {
  for (const map of mapping) {
    let fixedNumber: number | string = num / map.number;
    // We do this to prevent things like `1.0k`
    fixedNumber = Number.isInteger(fixedNumber) ? fixedNumber : fixedNumber.toFixed(precision);
    // If the number given is smaller than the current divider we have to continue to the next.
    // However, we wanna handle cases like `999` being transformed to `1k`
    if (map.number > num && Number(fixedNumber) < 1) continue;

    return `${fixedNumber}${space ? ' ' : ''}${
      upperCase ? map.postfix.toUpperCase() : map.postfix
    }`;
  }
  return num.toString();
}
