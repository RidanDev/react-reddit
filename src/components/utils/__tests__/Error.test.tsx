import React from 'react';
import { fireEvent } from 'react-testing-library';
import Error from '../Error';
import { renderWithTheme } from '../../../../test/utils';

describe('<Error />', () => {
  it('should set the message', () => {
    const { rerenderWithTheme, getByText } = renderWithTheme(<Error />);
    expect(getByText(/oops, something went wrong/i)).toBeInTheDocument();
    const message = 'Errrrrrror';
    rerenderWithTheme(<Error message={message} />);
    expect(getByText(message)).toBeInTheDocument();
  });

  it('should set the button message', () => {
    const { rerenderWithTheme, getByText } = renderWithTheme(<Error />);
    expect(getByText(/retry/i)).toBeInTheDocument();
    const buttonMessage = 'Button message';
    rerenderWithTheme(<Error buttonMessage={buttonMessage} />);
    expect(getByText(buttonMessage)).toBeInTheDocument();
  });

  it('should set the gutters', () => {
    const { container, rerenderWithTheme } = renderWithTheme(<Error />);
    expect(container.firstChild).toMatchSnapshot();
    rerenderWithTheme(<Error gutters />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should call `onButtonClick`', () => {
    const handleButtonClick = jest.fn();
    const { getByText } = renderWithTheme(<Error onButtonClick={handleButtonClick} />);
    fireEvent.click(getByText(/retry/i));
    expect(handleButtonClick).toHaveBeenCalledTimes(1);
  });
});
