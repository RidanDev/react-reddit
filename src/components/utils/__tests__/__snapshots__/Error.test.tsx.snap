// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Error /> should set the gutters 1`] = `
<div
  class="Hook-error"
  data-testid="error"
>
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root-161 MuiSvgIcon-colorDisabled-166"
    focusable="false"
    role="presentation"
    style="font-size: 80px;"
    viewBox="0 0 24 24"
  >
    <path
      d="M15.73 3H8.27L3 8.27v7.46L8.27 21h7.46L21 15.73V8.27L15.73 3zM12 17.3c-.72 0-1.3-.58-1.3-1.3 0-.72.58-1.3 1.3-1.3.72 0 1.3.58 1.3 1.3 0 .72-.58 1.3-1.3 1.3zm1-4.3h-2V7h2v6z"
    />
    <path
      d="M0 0h24v24H0z"
      fill="none"
    />
  </svg>
  <p
    class="MuiTypography-root-170 MuiTypography-body1-179 Hook-text"
  >
    Oops, something went wrong
  </p>
  <button
    class="MuiButtonBase-root-231 MuiButton-root-205 MuiButton-text-207 MuiButton-textPrimary-208 MuiButton-flat-210 MuiButton-flatPrimary-211"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label-206"
    >
      Retry
    </span>
    <span
      class="MuiTouchRipple-root-234"
    />
  </button>
</div>
`;

exports[`<Error /> should set the gutters 2`] = `
<div
  class="Hook-error Hook-gutters"
  data-testid="error"
>
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root-161 MuiSvgIcon-colorDisabled-166"
    focusable="false"
    role="presentation"
    style="font-size: 80px;"
    viewBox="0 0 24 24"
  >
    <path
      d="M15.73 3H8.27L3 8.27v7.46L8.27 21h7.46L21 15.73V8.27L15.73 3zM12 17.3c-.72 0-1.3-.58-1.3-1.3 0-.72.58-1.3 1.3-1.3.72 0 1.3.58 1.3 1.3 0 .72-.58 1.3-1.3 1.3zm1-4.3h-2V7h2v6z"
    />
    <path
      d="M0 0h24v24H0z"
      fill="none"
    />
  </svg>
  <p
    class="MuiTypography-root-170 MuiTypography-body1-179 Hook-text"
  >
    Oops, something went wrong
  </p>
  <button
    class="MuiButtonBase-root-231 MuiButton-root-205 MuiButton-text-207 MuiButton-textPrimary-208 MuiButton-flat-210 MuiButton-flatPrimary-211"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label-206"
    >
      Retry
    </span>
    <span
      class="MuiTouchRipple-root-234"
    />
  </button>
</div>
`;
